File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootZxart.java
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs;

import android.content.Context;
import android.content.res.Resources;
import android.net.Uri;
import android.util.SparseIntArray;

import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;

import app.zxtune.R;
import app.zxtune.fs.http.MultisourceHttpProvider;
import app.zxtune.fs.zxart.Author;
import app.zxtune.fs.zxart.CachingCatalog;
import app.zxtune.fs.zxart.Catalog;
import app.zxtune.fs.zxart.Identifier;
import app.zxtune.fs.zxart.Party;
import app.zxtune.fs.zxart.RemoteCatalog;
import app.zxtune.fs.zxart.Track;

public class VfsRootZxart extends StubObject implements VfsRoot {

  private static final String TAG = VfsRootZxart.class.getName();

  private final VfsObject parent;
  private final Context context;
  private final CachingCatalog catalog;
  private final GroupingDir[] groups;

  public VfsRootZxart(VfsObject parent, Context context, MultisourceHttpProvider http) {
    this.parent = parent;
    this.context = context;
    this.catalog = Catalog.create(context, http);
    this.groups = new GroupingDir[]{
        new AuthorsDir(),
        new PartiesDir(),
        new TopTracksDir()
    };
  }

  @Override
  public String getName() {
    return context.getString(R.string.vfs_zxart_root_name);
  }

  @Override
  public String getDescription() {
    return context.getString(R.string.vfs_zxart_root_description);
  }

  @Override
  public Uri getUri() {
    return Identifier.forRoot().build();
  }

  @Override
  @Nullable
  public VfsObject getParent() {
    return parent;
  }

  @Override
  public Object getExtension(String id) {
    if (VfsExtensions.SEARCH_ENGINE.equals(id)) {
      //assume search by authors from root
      return new AuthorsSearchEngine();
    } else if (VfsExtensions.ICON.equals(id)) {
      return R.drawable.ic_browser_vfs_zxart;
    } else {
      return super.getExtension(id);
    }
  }

  @Override
  public void enumerate(Visitor visitor) {
    for (GroupingDir group : groups) {
      visitor.onDir(group);
    }
  }

  @Override
  @Nullable
  public VfsObject resolve(Uri uri) {
    if (Identifier.isFromRoot(uri)) {
      final List<String> path = uri.getPathSegments();
      return resolve(uri, path);
    } else {
      return null;
    }
  }

  @Nullable
  private VfsObject resolve(Uri uri, List<String> path) {
    final String category = Identifier.findCategory(path);
    if (category == null) {
      return this;
    } else {
      for (GroupingDir group : groups) {
        if (category.equals(group.getPath())) {
          return group.resolve(uri, path);
        }
      }
    }
    return null;
  }

  private abstract static class GroupingDir extends StubObject implements VfsDir {

    @Override
    public Uri getUri() {
      return Identifier.forCategory(getPath()).build();
    }

    abstract String getPath();

    @Nullable
    abstract VfsObject resolve(Uri uri, List<String> path);
  }

  private class AuthorsDir extends GroupingDir {

    @Override
    public String getName() {
      return context.getString(R.string.vfs_zxart_authors_name);
    }

    @Override
    public VfsObject getParent() {
      return VfsRootZxart.this;
    }

    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.SEARCH_ENGINE.equals(id)) {
        return new AuthorsSearchEngine();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryAuthors(new Catalog.AuthorsVisitor() {
        @Override
        public void setCountHint(int hint) {
          visitor.onItemsCount(hint);
        }

        @Override
        public void accept(Author obj) {
          visitor.onDir(new AuthorDir(obj));
        }
      });
    }

    @Override
    public String getPath() {
      return Identifier.CATEGORY_AUTHORS;
    }

    @Override
    public VfsObject resolve(Uri uri, List<String> path) {
      // use plain resolving to avoid deep hierarchy
      // check most frequent cases first
      final Track track = Identifier.findAuthorTrack(uri, path);
      if (track != null) {
        return new AuthorTrackFile(uri, track);
      }
      final VfsObject dir = resolveAuthorDir(uri, path);
      return dir != null
          ? dir
          : this;
    }
  }

  @Nullable
  private VfsObject resolveAuthorDir(Uri uri, List<String> path) {
    final Author author = Identifier.findAuthor(uri, path);
    if (author == null) {
      return null;
    }
    final Integer year = Identifier.findAuthorYear(uri, path);
    return year != null
        ? new AuthorYearDir(author, year)
        : new AuthorDir(author);
  }

  private class AuthorDir extends StubObject implements VfsDir {

    private final Author author;

    AuthorDir(Author author) {
      this.author = author;
    }

    @Override
    public Uri getUri() {
      return Identifier.forAuthor(author).build();
    }

    @Override
    public String getName() {
      return author.getNickname();
    }

    @Override
    public String getDescription() {
      return author.getName();
    }

    @Override
    public VfsObject getParent() {
      return groups[0];// TODO
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      final SparseIntArray years = new SparseIntArray();

      catalog.queryAuthorTracks(author, new Catalog.TracksVisitor() {
        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public void accept(Track obj) {
          years.put(obj.getYear(), 1 + years.get(obj.getYear()));
        }
      });
      visitor.onItemsCount(years.size());
      for (int i = 0, lim = years.size(); i != lim; ++i) {
        // handle count when required
        visitor.onDir(new AuthorYearDir(author, years.keyAt(i)));
      }
    }
  }

  private class AuthorYearDir extends StubObject implements VfsDir {

    private final Author author;
    private final int year;

    AuthorYearDir(Author author, int year) {
      this.author = author;
      this.year = year;
    }

    @Override
    public Uri getUri() {
      return Identifier.forAuthor(author, year).build();
    }

    @Override
    public String getName() {
      return year != 0
          ? Integer.toString(year)
          : context.getString(R.string.vfs_zxart_unknown_year_name);
    }

    @Override
    public VfsObject getParent() {
      return new AuthorDir(author);
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryAuthorTracks(author, new Catalog.TracksVisitor() {
        @Override
        public void setCountHint(int size) {
          visitor.onItemsCount(size);
        }

        @Override
        public void accept(Track obj) {
          if (year == obj.getYear()) {
            final Uri uri = Identifier.forTrack(Identifier.forAuthor(author, year), obj).build();
            visitor.onFile(new AuthorTrackFile(uri, obj));
          }
        }
      });
    }
  }

  private class AuthorTrackFile extends BaseTrackFile {

    AuthorTrackFile(Uri uri, Track track) {
      super(uri, track);
    }

    @Override
    public VfsObject getParent() {
      final List<String> path = uri.getPathSegments();
      return resolveAuthorDir(uri, path);
    }
  }

  private class AuthorsSearchEngine implements VfsExtensions.SearchEngine {

    @Override
    public void find(String query, final Visitor visitor) throws IOException {
      catalog.findTracks(query, new Catalog.FoundTracksVisitor() {

        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public void accept(Author author, Track track) {
          final Uri uri = Identifier.forTrack(Identifier.forAuthor(author, track.getYear()), track).build();
          visitor.onFile(new AuthorTrackFile(uri, track));
        }
      });
    }
  }

  private class PartiesDir extends GroupingDir {

    @Override
    public String getName() {
      return context.getString(R.string.vfs_zxart_parties_name);
    }

    @Override
    public VfsObject getParent() {
      return VfsRootZxart.this;
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      final SparseIntArray years = new SparseIntArray();
      catalog.queryParties(new Catalog.PartiesVisitor() {
        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public void accept(Party obj) {
          years.put(obj.getYear(), 1 + years.get(obj.getYear()));
        }
      });
      visitor.onItemsCount(years.size());
      for (int i = 0, lim = years.size(); i != lim; ++i) {
        visitor.onDir(new PartyYearDir(years.keyAt(i)));
      }
    }

    @Override
    public String getPath() {
      return Identifier.CATEGORY_PARTIES;
    }

    @Override
    public VfsObject resolve(Uri uri, List<String> path) {
      final Track track = Identifier.findPartyTrack(uri, path);
      if (track != null) {
        return new PartyTrackFile(uri, track);
      }
      final VfsObject dir = resolvePartyDir(uri, path);
      return dir != null ? dir : this;
    }
  }

  @Nullable
  private VfsObject resolvePartyDir(Uri uri, List<String> path) {
    final Party party = Identifier.findParty(uri, path);
    if (party == null) {
      final Integer year = Identifier.findPartiesYear(uri, path);
      return year != null
          ? new PartyYearDir(year)
          : null;
    }
    final String compo = Identifier.findPartyCompo(uri, path);
    return compo != null
        ? new PartyCompoDir(party, compo)
        : new PartyDir(party);
  }

  private class PartyYearDir extends StubObject implements VfsDir {

    private final int year;

    PartyYearDir(int year) {
      this.year = year;
    }

    @Override
    public Uri getUri() {
      return Identifier.forPartiesYear(year).build();
    }

    @Override
    public String getName() {
      return Integer.toString(year);
    }

    @Override
    public VfsObject getParent() {
      return groups[1]; // TODO
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryParties(new Catalog.PartiesVisitor() {
        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public void accept(Party obj) {
          if (obj.getYear() == year) {
            visitor.onDir(new PartyDir(obj));
          }
        }
      });
    }
  }

  private class PartyDir extends StubObject implements VfsDir {

    private final Party party;

    PartyDir(Party party) {
      this.party = party;
    }

    @Override
    public Uri getUri() {
      return Identifier.forParty(party).build();
    }

    @Override
    public String getName() {
      return party.getName();
    }

    @Override
    public VfsObject getParent() {
      return new PartyYearDir(party.getYear());
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      final HashSet<String> compos = new HashSet<>();
      catalog.queryPartyTracks(party, new Catalog.TracksVisitor() {
        // TODO: remove
        @Override
        public void setCountHint(int size) {}

        @Override
        public void accept(Track obj) {
          compos.add(obj.getCompo());
        }
      });
      visitor.onItemsCount(compos.size());
      for (String compo : compos) {
        visitor.onDir(new PartyCompoDir(party, compo));
      }
    }
  }

  static class CompoName {
    private final String id;
    private final String name;

    private CompoName(String id, Context ctx) {
      this.id = id;
      this.name = getLocalizedName(ctx);
    }

    final String getId() {
      return id;
    }

    final String getName() {
      return name;
    }

    private String getLocalizedName(Context ctx) {
      final Resources res = ctx.getResources();
      final int resId = res.getIdentifier("vfs_zxart_compo_" + id, "string", ctx.getPackageName());
      return resId != 0
          ? res.getString(resId)
          : id;
    }
  }

  private static class PartyCompoTracksComparator implements Comparator<VfsObject> {
    @Override
    public int compare(VfsObject lh, VfsObject rh) {
      final int lhPlace = ((BaseTrackFile) lh).module.getPartyplace();
      final int rhPlace = ((BaseTrackFile) rh).module.getPartyplace();
      return Integer.compare(lhPlace, rhPlace);
    }

    static Comparator<VfsObject> instance() {
      return Holder.INSTANCE;
    }

    //onDemand holder idiom
    private static class Holder {
      static final PartyCompoTracksComparator INSTANCE = new PartyCompoTracksComparator();
    }
  }

  // custom ordering by party place
  private class PartyCompoDir extends StubObject implements VfsDir {

    private final Party party;
    private final CompoName compo;

    PartyCompoDir(Party party, String compo) {
      this.party = party;
      this.compo = new CompoName(compo, context);
    }

    @Override
    public Uri getUri() {
      return Identifier.forPartyCompo(party, compo.getId()).build();
    }

    @Override
    public String getName() {
      return compo.getName();
    }

    @Override
    public VfsObject getParent() {
      return new PartyDir(party);
    }

    @Nullable
    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.COMPARATOR.equals(id)) {
        return PartyCompoTracksComparator.instance();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryPartyTracks(party, new Catalog.TracksVisitor() {
        @Override
        public void setCountHint(int size) {
          visitor.onItemsCount(size);
        }

        @Override
        public void accept(Track obj) {
          if (compo.getId().equals(obj.getCompo())) {
            final Uri uri = Identifier.forTrack(Identifier.forPartyCompo(party, compo.getId()), obj)
                .build();
            visitor.onFile(new PartyTrackFile(uri, obj));
          }
        }
      });
    }
  }

  private class PartyTrackFile extends BaseTrackFile {

    PartyTrackFile(Uri uri, Track module) {
      super(uri, module);
    }

    @Override
    public VfsObject getParent() {
      final List<String> path = uri.getPathSegments();
      return resolvePartyDir(uri, path);
    }

    @Override
    public String getSize() {
      return Integer.toString(module.getPartyplace());
    }
  }

  private static class TopTracksComparator implements Comparator<VfsObject> {
    @Override
    public int compare(VfsObject lh, VfsObject rh) {
      // assume that votes can be compared in lexicographical order
      return -String.CASE_INSENSITIVE_ORDER.compare(((VfsFile) lh).getSize(),
          ((VfsFile) rh).getSize());
    }

    static Comparator<VfsObject> instance() {
      return Holder.INSTANCE;
    }

    //onDemand holder idiom
    private static class Holder {
      static final TopTracksComparator INSTANCE = new TopTracksComparator();
    }
  }

  // custom ordering by votes desc
  private class TopTracksDir extends GroupingDir {

    @Override
    public String getName() {
      return context.getString(R.string.vfs_zxart_toptracks_name);
    }

    @Override
    public String getDescription() {
      return context.getString(R.string.vfs_zxart_toptracks_description);
    }

    @Override
    public VfsObject getParent() {
      return VfsRootZxart.this;
    }

    @Nullable
    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.COMPARATOR.equals(id)) {
        return TopTracksComparator.instance();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryTopTracks(100, new Catalog.TracksVisitor() {
        @Override
        public void setCountHint(int size) {
          visitor.onItemsCount(size);
        }

        @Override
        public void accept(Track obj) {
          final Uri uri = Identifier.forTrack(Identifier.forCategory(Identifier.CATEGORY_TOP), obj)
              .build();
          visitor.onFile(new TopTrackFile(uri, obj));
        }
      });
    }

    @Override
    public String getPath() {
      return Identifier.CATEGORY_TOP;
    }

    @Override
    public VfsObject resolve(Uri uri, List<String> path) {
      final Track track = Identifier.findTopTrack(uri, path);
      return track != null
          ? new TopTrackFile(uri, track)
          : this;
    }

    private class TopTrackFile extends BaseTrackFile {

      TopTrackFile(Uri uri, Track module) {
        super(uri, module);
      }

      @Override
      public VfsObject getParent() {
        return TopTracksDir.this;
      }

      @Override
      public String getSize() {
        return module.getVotes();
      }
    }
  }

  private abstract static class BaseTrackFile extends StubObject implements VfsFile {

    final Uri uri;
    final Track module;

    BaseTrackFile(Uri uri, Track module) {
      this.uri = uri;
      this.module = module;
    }

    @Override
    public Uri getUri() {
      return uri;
    }

    @Override
    public String getName() {
      return uri.getLastPathSegment();
    }

    @Override
    public String getDescription() {
      return module.getTitle();
    }

    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.CACHE_PATH.equals(id)) {
        return Integer.toString(module.getId());
      } else if (VfsExtensions.DOWNLOAD_URIS.equals(id)) {
        return RemoteCatalog.getTrackUris(module.getId());
      } else if (VfsExtensions.SHARE_URL.equals(id)) {
        return getShareUrl();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public String getSize() {
      return module.getDuration();
    }

    private String getShareUrl() {
      return String.format(Locale.US, "https://zxart.ee/zxtune/action%%3aplay/tuneId%%3a%d",
          module.getId());
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootZxart.kt
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs

import android.content.Context
import android.net.Uri
import android.util.SparseIntArray
import androidx.annotation.StringRes
import androidx.core.util.keyIterator
import app.zxtune.R
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.fs.zxart.Author
import app.zxtune.fs.zxart.Catalog
import app.zxtune.fs.zxart.Identifier
import app.zxtune.fs.zxart.Party
import app.zxtune.fs.zxart.RemoteCatalog
import app.zxtune.fs.zxart.Track

class VfsRootZxart(
    override val parent: VfsObject, private val context: Context, http: MultisourceHttpProvider
) : StubObject(), VfsRoot {
    private val catalog = Catalog.create(context, http)
    private val groups = arrayOf(
        AuthorsDir(), PartiesDir(), TopTracksDir()
    )

    override val name
        get() = context.getString(R.string.vfs_zxart_root_name)
    override val description
        get() = context.getString(R.string.vfs_zxart_root_description)
    override val uri: Uri
        get() = Identifier.forRoot().build()

    override fun getExtension(id: String) = when (id) {
        //assume search by authors from root
        VfsExtensions.SEARCH_ENGINE -> AuthorsSearchEngine()
        VfsExtensions.ICON -> R.drawable.ic_browser_vfs_zxart
        else -> super.getExtension(id)
    }

    override fun enumerate(visitor: VfsDir.Visitor) = groups.forEach(visitor::onDir)

    override fun resolve(uri: Uri) = if (Identifier.isFromRoot(uri)) {
        resolve(uri, uri.pathSegments)
    } else {
        null
    }

    private fun resolve(uri: Uri, path: List<String>) =
        Identifier.findCategory(path)?.let { category ->
            groups.find { it.path == category }?.resolve(uri, path)
        } ?: this

    private abstract class GroupingDir(val path: String) : StubObject(), VfsDir {
        override val uri: Uri
            get() = id.build()

        abstract fun resolve(uri: Uri, path: List<String>): VfsObject?

        protected val id
            get() = Identifier.forCategory(path)
    }

    private inner class AuthorsDir : GroupingDir(Identifier.CATEGORY_AUTHORS) {
        override val name
            get() = context.getString(R.string.vfs_zxart_authors_name)
        override val parent
            get() = this@VfsRootZxart

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.SEARCH_ENGINE -> AuthorsSearchEngine()
            else -> super.getExtension(id)
        }

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryAuthors { obj ->
            visitor.onDir(AuthorDir(obj))
        }

        // use plain resolving to avoid deep hierarchy
        // check most frequent cases first
        override fun resolve(uri: Uri, path: List<String>) = Identifier.findAuthorTrack(
            uri, path
        )?.let { track ->
            AuthorTrackFile(uri, track)
        } ?: resolveAuthorDir(uri, path) ?: this
    }

    private fun resolveAuthorDir(uri: Uri, path: List<String>) =
        Identifier.findAuthor(uri, path)?.let { author ->
            Identifier.findAuthorYear(uri, path)?.let { year ->
                AuthorYearDir(author, year)
            } ?: AuthorDir(author)
        }

    private inner class AuthorDir(private val author: Author) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forAuthor(author).build()
        override val name
            get() = author.nickname
        override val description
            get() = author.name
        override val parent
            get() = groups[0] // TODO

        override fun enumerate(visitor: VfsDir.Visitor) = SparseIntArray().apply {
            catalog.queryAuthorTracks(author) { obj -> put(obj.year, 1 + get(obj.year)) }
        }.keyIterator().forEach { year ->
            // handle count when required
            visitor.onDir(AuthorYearDir(author, year))
        }
    }

    private inner class AuthorYearDir(private val author: Author, private val year: Int) :
        StubObject(), VfsDir {
        private val id
            get() = Identifier.forAuthor(author, year)

        override val uri: Uri
            get() = id.build()
        override val name
            get() = if (year != 0) year.toString() else context.getString(R.string.vfs_zxart_unknown_year_name)
        override val parent
            get() = AuthorDir(author)

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryAuthorTracks(author) { obj ->
            if (year == obj.year) {
                val uri = Identifier.forTrack(id, obj).build()
                visitor.onFile(AuthorTrackFile(uri, obj))
            }
        }
    }

    private inner class AuthorTrackFile(uri: Uri, track: Track) : BaseTrackFile(uri, track) {
        override val parent
            get() = resolveAuthorDir(uri, uri.pathSegments)
    }

    private inner class AuthorsSearchEngine : VfsExtensions.SearchEngine {
        override fun find(query: String, visitor: VfsExtensions.SearchEngine.Visitor) =
            catalog.findTracks(query) { author, track ->
                val uri =
                    Identifier.forTrack(Identifier.forAuthor(author, track.year), track).build()
                visitor.onFile(AuthorTrackFile(uri, track))
            }
    }

    private inner class PartiesDir : GroupingDir(Identifier.CATEGORY_PARTIES) {
        override val name
            get() = context.getString(R.string.vfs_zxart_parties_name)
        override val parent
            get() = this@VfsRootZxart

        override fun enumerate(visitor: VfsDir.Visitor) = SparseIntArray().apply {
            catalog.queryParties { obj -> put(obj.year, 1 + get(obj.year)) }
        }.keyIterator().forEach { year ->
            // handle count when required
            visitor.onDir(PartyYearDir(year))
        }

        override fun resolve(uri: Uri, path: List<String>) =
            Identifier.findPartyTrack(uri, path)?.let { track ->
                PartyTrackFile(uri, track)
            } ?: resolvePartyDir(uri, path) ?: this
    }

    private fun resolvePartyDir(uri: Uri, path: List<String>) =
        Identifier.findParty(uri, path)?.let { party ->
            Identifier.findPartyCompo(uri, path)?.let { compo ->
                PartyCompoDir(party, compo)
            } ?: PartyDir(party)
        } ?: Identifier.findPartiesYear(uri, path)?.let { year ->
            PartyYearDir(year)
        }

    private inner class PartyYearDir(private val year: Int) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forPartiesYear(year).build()
        override val name
            get() = year.toString()
        override val parent
            get() = groups[1] // TODO

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryParties { obj ->
            if (obj.year == year) {
                visitor.onDir(PartyDir(obj))
            }
        }
    }

    private inner class PartyDir(private val party: Party) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forParty(party).build()
        override val name
            get() = party.name
        override val parent
            get() = PartyYearDir(party.year)

        override fun enumerate(visitor: VfsDir.Visitor) = HashSet<String>().apply {
            catalog.queryPartyTracks(party) { obj -> add(obj.compo) }
        }.forEach { compo ->
            visitor.onDir(PartyCompoDir(party, compo))
        }
    }

    enum class LocalizedCompoName(@StringRes val titleRes: Int) {
        UNKNOWN(R.string.vfs_zxart_compo_unknown), STANDARD(R.string.vfs_zxart_compo_standard), AY(R.string.vfs_zxart_compo_ay), BEEPER(
            R.string.vfs_zxart_compo_beeper
        ),
        COPYAY(R.string.vfs_zxart_compo_copyay), NOCOPYAY(R.string.vfs_zxart_compo_nocopyay), REALTIME(
            R.string.vfs_zxart_compo_realtime
        ),
        REALTIMEAY(R.string.vfs_zxart_compo_realtimeay), REALTIMEBEEPER(R.string.vfs_zxart_compo_realtimebeeper), OUT(
            R.string.vfs_zxart_compo_out
        ),
        WILD(R.string.vfs_zxart_compo_wild), EXPERIMENTAL(R.string.vfs_zxart_compo_experimental), OLDSCHOOL(
            R.string.vfs_zxart_compo_oldschool
        ),
        MAINSTREAM(R.string.vfs_zxart_compo_mainstream), PROGRESSIVE(R.string.vfs_zxart_compo_progressive), TS(
            R.string.vfs_zxart_compo_ts
        ),
        TSFM(R.string.vfs_zxart_compo_tsfm), RELATED(R.string.vfs_zxart_compo_related);

        companion object {
            fun get(id: String, ctx: Context) = id.uppercase().let { name ->
                entries.find { it.name == name }
            }?.run {
                ctx.getString(titleRes)
            } ?: id
        }
    }

    private object PartyCompoTracksComparator : Comparator<VfsObject> {
        override fun compare(lh: VfsObject, rh: VfsObject): Int {
            val lhPlace = (lh as BaseTrackFile).module.partyplace
            val rhPlace = (rh as BaseTrackFile).module.partyplace
            return lhPlace.compareTo(rhPlace)
        }
    }

    // custom ordering by party place
    private inner class PartyCompoDir(private val party: Party, private val compo: String) :
        StubObject(), VfsDir {
        private val id
            get() = Identifier.forPartyCompo(party, compo)

        override val uri: Uri
            get() = id.build()
        override val name
            get() = LocalizedCompoName.get(compo, context)
        override val parent
            get() = PartyDir(party)

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.COMPARATOR -> PartyCompoTracksComparator
            else -> super.getExtension(id)
        }

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryPartyTracks(party) { obj ->
            if (obj.compo == compo) {
                val uri = Identifier.forTrack(id, obj).build()
                visitor.onFile(PartyTrackFile(uri, obj))
            }
        }
    }

    private inner class PartyTrackFile(uri: Uri, module: Track) : BaseTrackFile(uri, module) {
        override val parent
            get() = resolvePartyDir(uri, uri.pathSegments)
        override val size
            get() = module.partyplace.toString()
    }

    private object TopTracksComparator : Comparator<VfsObject> {
        override fun compare(lh: VfsObject, rh: VfsObject): Int {
            // assume that votes can be compared in lexicographical order
            return -java.lang.String.CASE_INSENSITIVE_ORDER.compare(
                (lh as VfsFile).size, (rh as VfsFile).size
            )
        }
    }

    // custom ordering by votes desc
    private inner class TopTracksDir : GroupingDir(Identifier.CATEGORY_TOP) {
        override val name
            get() = context.getString(R.string.vfs_zxart_toptracks_name)
        override val description
            get() = context.getString(R.string.vfs_zxart_toptracks_description)
        override val parent
            get() = this@VfsRootZxart

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.COMPARATOR -> TopTracksComparator
            else -> super.getExtension(id)
        }

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryTopTracks(100) { obj ->
            val uri = Identifier.forTrack(id, obj).build()
            visitor.onFile(TopTrackFile(uri, obj))
        }

        override fun resolve(uri: Uri, path: List<String>) =
            Identifier.findTopTrack(uri, path)?.let { track ->
                TopTrackFile(uri, track)
            } ?: this

        private inner class TopTrackFile(uri: Uri, module: Track) : BaseTrackFile(uri, module) {
            override val parent
                get() = this@TopTracksDir
            override val size
                get() = module.votes
        }
    }

    private abstract class BaseTrackFile(override val uri: Uri, val module: Track) : StubObject(),
        VfsFile {
        override val name
            get() = uri.lastPathSegment ?: ""

        override val description
            get() = module.title

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.CACHE_PATH -> module.id.toString()
            VfsExtensions.DOWNLOAD_URIS -> RemoteCatalog.getTrackUris(module.id)
            VfsExtensions.SHARE_URL -> shareUrl
            else -> super.getExtension(id)
        }

        override val size
            get() = module.duration

        private val shareUrl
            get() = "https://zxart.ee/zxtune/action%%3aplay/tuneId%%3a${module.id}"
    }
}
